[fx] Pass manager

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79879


PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

Differential Revision: [D37459319](https://our.internmc.facebook.com/intern/diff/D37459319)

angelayi committed:
PassManager Implementation

[ghstack-poisoned]

angelayi committed:
Update on "PassManager Implementation"




[ghstack-poisoned]

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

[ghstack-poisoned]

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

[ghstack-poisoned]

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

[ghstack-poisoned]

SherlockNoMad commented on torch/fx/passes/infra/pass_manager.py:
```diff
@@ -0,0 +1,169 @@
+            after each pass
+    """
+
+    passes: List[Callable] = []
```
We want to be explicit that this is a Callable[[GraphModule], None]...

SherlockNoMad commented on torch/fx/passes/infra/pass_manager.py:
```diff
@@ -0,0 +1,169 @@
+        self.constraints.append(constraint)
+        self._validated = False
+
+    def validate_constraints(self):
```
I think we should find a valid pass application order according to the constrain.

SherlockNoMad replied:
The collection of constrains forms a dependency graph, and we just need to do a topological sort to find a valid traversal order.

SherlockNoMad commented on torch/fx/passes/infra/pass_manager.py:
```diff
@@ -0,0 +1,169 @@
+        `steps` number of times.
+        """
+        # Check constraints
+        self.validate_constraints()
```
let's solve_constrains() here

SherlockNoMad commented on torch/fx/passes/infra/pass_manager.py:
```diff
@@ -0,0 +1,169 @@
+
+            graph_module.recompile()
+
+            # If the graph no longer changes, then we can stop running these passes
```
I wonder if this is a reliable way to check if graph has been modified.... 
I think we PassBase should return a boolean, say "modified", to indicate if it actually took effect on modifying the graph. 
This "modified" return field could also be used to collect Pass Statistics.

angelayi replied:
I noticed that ONNX does this too (have the pass base return a `modified` boolean), but I wasn't sure if it would be too much work for users to have to manage this information when implementing their pass.

However, having the user track this state does seem like the most accurate way to check if the graph has been modified. Right now I'm checking if the `graph_module.code` values are the same -- justification being that `graph_module.recompile()` recompiles the `code` based on the `graph`, so I assume that if the graph changes then the code will also change, and if the graph doesn't change then the code will also not change. I'm not positive this is a full-proof solution though.

What do you think?

zhxchen17 replied:
I saw that in Torchscript passes, and it seems to me that users usually wouldn't have difficulty returning this changing status to framework. I think in the long term it might be more flexible if users can just return a PassResult type and `changed` can be one of the data fields.

SherlockNoMad commented on torch/fx/passes/infra/pass_manager.py:
```diff
@@ -0,0 +1,169 @@
+
+        setattr(self, "check", check)  # noqa: B010
+
+    def check(self, graph_module: GraphModule) -> None:
```
what happens if check failed? 
Do we raise error and crash? or log warning and continue?

angelayi replied:
I was thinking this could be up whatever the users decide since they're implementing the `check` function. What do you think?

SherlockNoMad commented on test/test_fx.py:
```diff
@@ -42,6 +42,7 @@
 from fx.test_dce_pass import TestDCE  # noqa: F401
 from fx.test_fx_const_fold import TestConstFold  # noqa: F401
 from fx.test_fx_param_shape_control_flow import TestConstParamShapeInControlFlow  # noqa: F401
+from fx.test_pass_infra import TestPassManager  # noqa: F401
```
do you need this line?

angelayi replied:
I think I do in order to run `python test/test_fx.py TestPassManager`

SherlockNoMad replied:
ic.

SherlockNoMad commented on torch/fx/passes/pass_manager.py:
```diff
@@ -1,188 +0,0 @@
-
-# for callables which modify object inplace and return something other than
-# the object on which they act
-def inplace_wrapper(fn: Callable) -> Callable:
```
this is a BC breaking change. 
fbcode has dependency on this function https://www.internalfb.com/code/fbsource/[d3381e07aff67fc001520c7e2018816b27cc8993]/fbcode/deeplearning/trt/torch_tensorrt/py/torch_tensorrt/fx/passes/lower_pass_manager_builder.py?lines=6%2C98%2C110

Let's talk offline. we will need a special workflow for merging this PR.

SherlockNoMad replied:
same for other changes that removed code.

SherlockNoMad commented on test/fx/test_pass_infra.py:
```diff
@@ -0,0 +1,89 @@
+        if node.op == "call_function" and node.target == torch.add:
+            node.target = torch.mul
+
+def replace_mul_with_sub_pass(gm):
```
replace_mul_with_div_pass?

SherlockNoMad commented on test/fx/test_pass_infra.py:
```diff
@@ -0,0 +1,89 @@
+            pass
+
+        pm = PassManager()
+        self.assertRaises(TypeError, pm.add_checks, check_bad_args)
```
It's more intuitive to write 

with self.assertRaise(...): 
     pm.add_checks(check_bad_args)

SherlockNoMad reviewed (changes requested):
Thanks a lot for implementing this! 
Let's chat offline about how to proceed with BC breaking change.

SherlockNoMad commented:
I just learnt the workflow for handling fbcode breaking change

1. Import the oss PR into fbcode as a diff
2. fix internal breaking change in the diff
3. land the diff, and it will also trigger the merge on oss PR

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37459319).

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

Differential Revision: [D37459319](https://our.internmc.facebook.com/intern/diff/D37459319)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37459319).

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

Differential Revision: [D37459319](https://our.internmc.facebook.com/intern/diff/D37459319)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37459319).

angelayi committed:
Update on "[fx] Pass manager"



PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

Differential Revision: [D37459319](https://our.internmc.facebook.com/intern/diff/D37459319)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37459319).

angelayi commented:
Sorry! Moving changes to https://github.com/pytorch/pytorch/pull/80531 since I messed up some dependencies.
